### ISSUE TITLE
cancellation too presumptive?

### ISSUE THREAD

**Issue by aft-luke** [yelouafi/redux-saga#53]
Just some food for thought...

> Nevertheless, a warning is logged into the console in case a cancelled task omitted to handle a cancellation exception.

I have quite a few instances where one saga spawns a fetcher function, and that fetcher func then invokes the actual function doing the xhr. This is to keep functions lean and "doing one thing". Maybe I want to transform the parameters or data before/after fetching, or maybe I want to check status codes and validate returned values versus nulls, etc.

```javascript
    function doFetch(params) {
        yield xhr(url, params)
    }

    function* fetch(params) {
        const { statusCode, body } = yield call(doFetch, params)
        
        if (statusCode > 499) {
            return null
        }

        const data = body.someData

        // ... do some transformations on the data

        return {
            transformedData
        }
    }

    function* fetchSomeStuff() {
        yield put(START_FETCH)

        // ...maybe configure some parameters

        const results = yield call(fetch, params)

        // ...do some data transforms on results

        yield put(FETCH_SUCCESS, results)
    }
```

When I want to sure up this code with some error handling, in order to avoid a `console.warn` by redux-saga I have to put in a try/catch for every step along the saga. This is fairly inconvenient to have to handle a thrown saga cancelled exception at every invocation, particularly because I only want to use the catch to "cleanup" in one of the cases:

```javascript
    function doFetch(params) {
        let req
        try {
            return (yield req = xhr(url, params))
        } catch (e) {
            if (e instanceof SagaCancellationException) {
                req.abort()
            } else {
                throw e
            }
        }
    }

    function* fetch(params) {
        // doesn't have to handle errors, since there's nothing to do and the
        // caller handles anything that might be thrown... 
        try {
            // ... same code as above
        } catch (e) {
            // there's nothing to do here except stop the console.warn() from showing up
        }
    }

    function* fetchSomeStuff() {
        try {
            yield put(START_FETCH)

            // ...maybe configure some parameters

            const results = yield call(fetch, params)

            // ...do some data transforms on results

            yield put(FETCH_SUCCESS, results)
        } catch (e) {
            // capture SagaCancellationException, but throw any js errors
            if (!e instanceof io.SagaCancellationException) {
                console.error(e)
            }
        }
    }
```

**Comment by yelouafi** [comment]
Not sure I understand the issue. Saga cancellation exceptions are only thrown when yielding a cancel effect. But I dont see any cancel effect in the code above

**Comment by aft-luke** [comment]
Sorry, I left out an important piece - the root saga invoking this chain.

```javascript
    function* root() {
        let process
        while (true) {
            yield take(SOME_TRIGGER)
            if (process && process.isRunning()) {
                yield cancel(process)
            }

            process = yield fork(fetchSomeStuff)
        }
    }
```

I use this pattern when the user can click on a list of items, and each click will trigger a data fetch for that item. If the user changes his mind and clicks on a different item while one is already loading, we can cancel the first data fetch.

It's not a huge ordeal, I'm just pointing out that when `yield cancel(process)` is triggered, it will throw `SagaCancellationException` in each branch of the `fetchSomeStuff`- which is fine, but it might be too ambitious to `console.warn` in each branch that doesn't explicitly handle that exception. I don't want to have to write a `try/catch` for every level of my saga if only one (bottom-most) or two (bottom most and top most) levels actually do anything after the exception.

**Comment by yelouafi** [comment]
>but it might be too ambitious to console.warn in each branch that doesn't explicitly handle that exception

the warning is printed just to remind the developer in case he forgot to handle the cancellation. Which can lead to some mysterious bugs (inconsistent state)

**Comment by yelouafi** [comment]
I'm thinking of making this an opt-in behavior. Provide some config argument, so the developer can choose to disable the console warnings

**Comment by aft-luke** [comment]
I would :+1:  that... perhaps just the `NODE_ENV="production"` convention. I have used that with something like [envify](https://github.com/zertosh/loose-envify) in the past and it works well.

**Comment by yelouafi** [comment]
>perhaps just the NODE_ENV="production" convention. I have used that with something like envify in the past and it works well.

Agree; I think it's a better idea

**Comment by gaearon** [comment]
Note that you'll want to make sure error behavior is exactly the same between prod and dev versions. If anything throws, it should throw in both cases. Logging a warning only in dev is fine though.

**Comment by aft-luke** [comment]
Looks like this was taken care of in https://github.com/yelouafi/redux-saga/blob/no-promise/src/proc.js#L79-L81

**Comment by yelouafi** [comment]
@aft-luke yes. Included with the new 0.6.0
